calculator/src/main/java/com/market/banica/calculator/service/JMXServiceImpl.java

```diff
@@ -0,0 +1,299 @@
+import org.springframework.jmx.export.annotation.ManagedResource;
+import org.springframework.stereotype.Service;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Locale;
+import java.util.Map;
+
+@EnableMBeanExport
+@ManagedResource
+@Service
+@RequiredArgsConstructor
+public class JMXServiceImpl implements JMXService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(JMXServiceImpl.class);
+
+    private final RecipesBase recipesBase;
+    private final ProductService productService;
+
+    @Override
+    @ManagedOperation
+    public Map<String,Product> getDatabase(){
+      return   recipesBase.getDatabase();
+    }
+
+    @Override
+    @ManagedOperation
+    public void createProduct(String newProductName, String unitOfMeasure, Map<String, Integer> ingredients) {
+        LOGGER.debug("JMX server impl: In createProduct method");
+        LOGGER.info("CreateProduct called from JMX server with parameters newProductName {},unitOfMeasure {}," +
+                        " ingredients {} and ingredients {}", newProductName, unitOfMeasure,
+                ingredients, ingredients);
+        Product product = retrieveProductFromDatabase(newProductName);
+
+        if (product == null) {
+            createNewProduct(newProductName, unitOfMeasure, ingredients);
+
+            LOGGER.info("New product created from JMX server with product name {} and unit of measure {}"
+                    , newProductName, unitOfMeasure);
+
+            productService.createBackUp();
+        } else {
+
+            LOGGER.error("Product with name {} already exists", newProductName);
+            throw new IllegalArgumentException("Product with this name already exists");
+        }
+    }
+
+    @Override
+    @ManagedOperation
+    public void addIngredient(String recipeName, String ingredientName, String quantityAsString) {
+        LOGGER.debug("JMX server impl: In addIngredient method");
+        LOGGER.info("AddIngredient called from JMX server with parameters recipeName {},ingredientName {} and quantityAsString {}",
+                recipeName, ingredientName, quantityAsString);
+
+        Product ingredient = retrieveProductFromDatabase(ingredientName);
+        Product recipe = retrieveProductFromDatabase(recipeName);
+
+        validateProductExist(ingredientName, ingredient);
+        validateProductExist(recipeName, recipe);
+
+        int quantity = getValueAsInt(quantityAsString);
+
+        recipe.getIngredients().put(ingredientName,quantity);
+
+        productService.createBackUp();
+
+        LOGGER.info("Ingredient added from JMX server for recipe {} and ingredient {} with value {}"
+                , recipeName, ingredientName, quantityAsString);
+    }
+
+    @Override
+    @ManagedOperation
+    public void setProductQuantity(String recipeName, String ingredientName, String newValue) {
+        LOGGER.debug("JMX server impl: In setProductQuantity method");
```

Comment:
newValue is not a good variable name, please be more specific.

Reply:
Done